osu.Game/OsuGame.cs

```diff
@@ -742,24 +744,34 @@ public void PresentScore(IScoreInfo score, ScorePresentType presentType = ScoreP
         {
             Logger.Log($"Beginning {nameof(PresentScore)} with score {score}");
 
-            var databasedScore = ScoreManager.GetScore(score);
+            Score databasedScore = null;
 
-            if (databasedScore == null) return;
-
-            if (databasedScore.Replay == null)
+            try
+            {
+                databasedScore = ScoreManager.GetScore(score);
+            }
+            catch (LegacyScoreDecoder.BeatmapNotFoundException notFound)
             {
-                Logger.Log("The loaded score has no replay data.", LoggingTarget.Information);
+                Logger.Log("The replay cannot be played because the beatmap is missing.", LoggingTarget.Information);
+
+                var req = new GetBeatmapRequest(new BeatmapInfo { MD5Hash = notFound.Hash });
+                req.Success += res => Notifications.Post(new MissingBeatmapNotification(res, notFound.Hash, null));
+                API.Queue(req);
+
                 return;
             }
 
-            var databasedBeatmap = BeatmapManager.QueryBeatmap(b => b.ID == databasedScore.ScoreInfo.BeatmapInfo.ID);
+            if (databasedScore == null) return;
 
-            if (databasedBeatmap == null)
+            if (databasedScore.Replay == null)
             {
-                Logger.Log("Tried to load a score for a beatmap we don't have!", LoggingTarget.Information);
+                Logger.Log("The loaded score has no replay data.", LoggingTarget.Information, LogLevel.Important);
                 return;
             }
```

Comment:
I deleted this entire block because in light of all of the actions that `ScoreManager.GetScore()` does, namely

https://github.com/ppy/osu/blob/4a95973b6e074f6683131c6a7e3f7defc82b5ca1/osu.Game/Scoring/ScoreManager.cs#L89-L112

and

https://github.com/ppy/osu/blob/4a95973b6e074f6683131c6a7e3f7defc82b5ca1/osu.Game/Scoring/Legacy/DatabasedLegacyScoreDecoder.cs#L27
https://github.com/ppy/osu/blob/4a95973b6e074f6683131c6a7e3f7defc82b5ca1/osu.Game/Scoring/Legacy/LegacyScoreDecoder.cs#L66-L69

I find it quite unlikely that this check isn't just completely dead code and a redundant refetch, but maybe tests will prove me wrong or something.